### PR TITLE
X.509 certificates are generated dynamically during tests.

### PR DESCRIPTION
**Pull request by lukasz-soszynski-eliatra**
Signed-off-by: Lukasz Soszynski <lukasz.soszynski@eliatra.com>

### Description
[Describe what this change achieves]
* Category Enhancement
* Test framework generates dynamically X.509 certificates which are used in the course of tests
* Hardcoded certificates were used

### Issues Resolved
[List any issues this PR will resolve]

Is this a backport? If so, please add backport PR # and/or commits #

### Testing
This is a part of the test framework. I nodes during tests start correctly and then tests are executed correctly this means that communication between OpenSearch nodes with generated certificates works correctly.

### Check List
- [ ] New functionality includes testing
- [ ] New functionality has been documented
- [ ] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by peternied — APPROVED** [review]
This is a great contribution.

File: build.gradle [hunk 1]
```diff
@@ -298,7 +299,7 @@ dependencies {
     implementation 'com.google.guava:guava:30.0-jre'
     implementation 'org.greenrobot:eventbus:3.2.0'
     implementation 'commons-cli:commons-cli:1.3.1'
-    implementation 'org.bouncycastle:bcprov-jdk15on:1.67'
+    implementation "org.bouncycastle:bcprov-jdk18on:${bouncycastle_version}"
```
Comment (cwperks) [inline build.gradle:302]: Should we be matched with core for the bouncy castle version? Core has `bouncycastle` listed in `buildSrc/version.properties`

You can refer to these properties like `${versions.bouncycastle}`

File: src/integrationTest/java/org/opensearch/test/framework/certificate/AlgorithmKit.java [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+* Please see {@link #getSignatureAlgorithmName()}
+*/
+class AlgorithmKit {
+
+	private static final Logger log = LogManager.getLogger(AlgorithmKit.class);
+	public static final String SIGNATURE_ALGORITHM_SHA_256_WITH_RSA = "SHA256withRSA";
+	public static final String SIGNATURE_ALGORITHM_SHA_256_WITH_ECDSA = "SHA256withECDSA";
+
+	private final String signatureAlgorithmName;
+	private final Supplier<KeyPair> keyPariSupplier;
```
Comment (cwperks) [inline src/integrationTest/java/org/opensearch/test/framework/certificate/AlgorithmKit.java:56]: nit: keyPairSupplier*

File: src/integrationTest/java/org/opensearch/test/framework/certificate/TestCertificates.java [hunk 3]
```diff
@@ -32,46 +32,146 @@
...
+		this.caCertificate = createCaCertificate();
+		this.nodeCertificates = IntStream.range(0, MAX_NUMBER_OF_NODE_CERTIFICATES)
+			.mapToObj(this::createNodeCertificate)
+			.collect(Collectors.toList());
+		this.adminCertificate = createAdminCertificate();
+	}
+
+
+	private CertificateData createCaCertificate() {
+		CertificateMetadata metadata = CertificateMetadata.basicMetadata(CA_SUBJECT, CERTIFICATE_VALIDITY_DAYS)
```
Comment (cwperks) [inline src/integrationTest/java/org/opensearch/test/framework/certificate/TestCertificates.java:75]: 🙌  This is great!

File: src/integrationTest/java/org/opensearch/test/framework/certificate/TestCertificates.java [hunk 4]
```diff
@@ -32,46 +32,146 @@
...
-* This will be replaced by classes
-* that can generate certificates on the fly.
+* It provides TLS certificates required in test cases. The certificates are generated during process of creation objects of the class.
+* The class exposes method which can be used to write certificates and private keys in temporally files.
 */
 public class TestCertificates {
 
+	public static final Integer MAX_NUMBER_OF_NODE_CERTIFICATES = 3;
+
+	private static final String CA_SUBJECT = "DC=com,DC=example,O=Example Com Inc.,OU=Example Com Inc. Root CA,CN=Example Com Inc. Root CA";
```
Comment (cwperks) [inline src/integrationTest/java/org/opensearch/test/framework/certificate/TestCertificates.java:55]: In some of our tests I see that we have intermediate signing certificates: https://github.com/opensearch-project/security/blob/main/src/test/java/org/opensearch/security/ssl/SecuritySSLCertsInfoActionTests.java#L36

With subject `CN=Example Com Inc. Signing CA,OU=Example Com Inc. Signing CA,O=Example Com Inc.,DC=example,DC=com`.

Can we create intermediate signing CAs with this framework as well?

**Review by cwperks — APPROVED** [review]
Thank you @lukasz-soszynski-eliatra!

File: build.gradle (hunk 1 above)
Comment (lukasz-soszynski-eliatra) [reply build.gradle:302]: I will try to use a common version. Initially, I thought that the library with version postfix `jdk18on` should be used with Java 11. But this probably is not the case...

File: src/integrationTest/java/org/opensearch/test/framework/certificate/TestCertificates.java (hunk 4 above)
Comment (lukasz-soszynski-eliatra) [reply src/integrationTest/java/org/opensearch/test/framework/certificate/TestCertificates.java:55]: To be honest, I did not test this, but I assume that it should be possible. The intermediate certificate is much like a root certificate (basic constrain CA:true, and proper key usage, etc.) but additionally signed with the private key associated with the parent certificate.

**Comment by lukasz-soszynski-eliatra** [comment]
Thanks for your review :)

File: build.gradle (hunk 1 above)
Comment (lukasz-soszynski-eliatra) [reply build.gradle]: Done, common version of Bouncy Castle used

File: src/integrationTest/java/org/opensearch/test/framework/certificate/AlgorithmKit.java (hunk 2 above)
Comment (lukasz-soszynski-eliatra) [reply src/integrationTest/java/org/opensearch/test/framework/certificate/AlgorithmKit.java]: done

**Review by cwperks — APPROVED** [review]

**Review by peternied — APPROVED** [review]

**Pull request merged**